### ISSUE TITLE
undefined symbol _imp_select

### ISSUE THREAD

**Issue by roboLoki** [jtv/libpqxx#186]
HI all
In the stable branch I installed libpqxx for 11.3 postgres. Everything went cool but I get this message. It says  (): >>> referenced by libpqxx.a(connection_base.o):(_ZN12_GLOBAL__N_17wait_fdEibP7timeval)
So I googled and that told me it is some symbol in the .a file that is conflicting. It said use a dll if you can. I cannot because the following did not yield a dll…
./configure --enable-shared=yes --enable-static=no --disable-documentation
Am I correct that with mingw because there is a symbol that I am not able to use libpqxx in mingw?  I will include the test code for your review:
`#include <iostream>
#include <pqxx/pqxx> 
#include "libpq-fe.h"

using namespace std;
using namespace pqxx;

int main(int argc, char* argv[]) {
   try {
      connection C("dbname = testdb user = apophis password = foo   \
      hostaddr = 127.0.0.1 port = 5432");
      if (C.is_open()) {
         cout << "Opened database successfully: " << C.dbname() << endl;
      } else {
         cout << "Can't open database" << endl;
         return 1;
      }
      C.disconnect ();
   } catch (const std::exception &e) {
      cerr << e.what() << std::endl;
      return 1;
   }
}`

thnx so much for any assistance you can provide. roboloki

**Comment by tt4g** [comment]
I want some more information.
When will you see that error message? (e.g. compilation time, execute program, etc ...)

If you get an error at build time, you will also need the operating system and compiler information.

> In the stable branch I installed libpqxx for 11.3 postgres

Are you using libpqxx 6.4.4?

**Comment by tt4g** [comment]
If you get an error message at running program, I point out that you may have misspecified the connection parameters  (it may be the wrong answer because the source code is not well-formed).

In this case you have to remove the space before and after `=` which separates the connection parameter.

```diff
- connection C("dbname = testdb user = apophis password = foo 
- hostaddr = 127.0.0.1 port = 5432");
+connection C("dbname=testdb user=apophis password=foo hostaddr=127.0.0.1 port=5432");
```

**Comment by roboLoki** [comment]
Hi-

Win 10 home version 1809.
I'm using Ultimate++ IDE with mingw64 recently updated with pacman.
When I hit the yellow lightning bolt to link.
The code compiles fine when I just hit compile.
My build method looks like this:
x86_64-w64-mingw32-g++.exe
--std=c++14 Wl,--stack,20000000   -lpqxx -lpq
-msse2 -D__CRT__NO_INLINE
I am using the stable master branch off git that says 5.1.
from https://github.com/jtv/libpqxx

I hope that info helps.
thnx roboloki

**Comment by roboLoki** [comment]
Hi jtv/Libpqxx
Removing spaces left same error msg.
Thnx roboloki

**Comment by tt4g** [comment]
The libpqxx `configure` script only supports Unix-like systems.
MinGW may not be supported.

Can you try building with `cmake` ?
The procedure is described in README.md.

**Comment by tt4g** [comment]
Sorry. I had overlooked.
Here is the procedure for MinGW: https://github.com/jtv/libpqxx/blob/6.4.4/win32/INSTALL.txt

You may want to try this before `cmake`.

**Comment by roboLoki** [comment]
Hi jtv-

I think you put me on to something. I saw this reference to ws2_32 in
someone's google post regarding my error and didn't know how to do it. Now
I see how to do it in the link you gave me, and I just have to find the
package that has that file and use the install instructions from your link. I
am trying this to diagnose for i.e. to see if a file is installed: $
pkgfile g++.exe
mingw32/mingw-w64-i686-gcc
mingw64/mingw-w64-x86_64-gcc
msys/gcc but pkgfile ws2_32.dll or pkgfile ws2_32.a didn't return
anything. tomorrow
I will search out the package in mingw and run with that. I think if I can
find that file and install its package it should work. I'm at gmt -4, and I
need some sleep for a day back at work after a vacation or my boss will be
none too happy. thnx roboloki!

**Comment by tt4g** [comment]
`ws2_32` is a library for socket communication in Windows.

As MinGW seems to use `ws2_32` for socket communication, I think that additional installation is unnecessary: [Ws2_32.lib is not on my system - C++ Forum](http://www.cplusplus.com/forum/beginner/240684/)

**Comment by roboLoki** [comment]
Hey tt4g- You rock. 
You need the ws2_32.dll from windows system32 or you get an error msg when linking.
Thnx. I got it with this command-line:
$ x86_64-w64-mingw32-g++.exe -I. -Ic:/pgsql/include -L. -Lc:/pgsql/lib  -Lc:/msys64/mingw64/lib ws2_32.dll  pgtest.cpp -lpqxx -lpq  -o wott.exe
$ ./wott.exe
Opened database successfully: testdb
I didn't get U++ up and running with this issue. I left them some notes. I am not sure if they will check it out. It may be something I am doing wrong. Not sure.
Have a very cool night. Thnx a lot. roboloki

**Comment by jtv** [comment]
I'm sorry for not stepping in earlier — I'm not entirely well at the moment and it's a bit distracting.  But I'd like to make sure that other people don't hit the same problem, so I can hopefully fix it!

**Comment by roboLoki** [comment]
Hi-
There isn't too much to say except you needed a .pc config file, so that
the configure command in mingw can locate your include and lib files. The
rest is reading the mingw notes specific to that platform. I didn't realize
at first I needed to link against an extra library, but that is in the
install notes covered in the mingw section. I didn't know at first the
extra DLL file is in system32 dir in Windows. I looked around in the
package manager for it. I can't think of anything else. jtv pointed me in
the right direction. Thnx for the cool help.  Roboloki

**Comment by jtv** [comment]
@roboLoki I think @tt4g was the one who pointed you in the right direction!

I do have a `libpqxx.pc` in the main source directory.  Is that the one you needed?

**Comment by roboLoki** [comment]
Hi-
The configure script couldnt find that, and I didn't know it was available,
so I made my own copy based on what others did. The configure printed out
very good error messages, and that is what tipped me off. I don't know why
configure didn't see the one you made. Could have been different install
paths.
Thnx roboloki

**Comment by jtv** [comment]
Maybe there was something different inside the file itself..?

**Comment by tt4g** [comment]
@jtv Since there is a configure script in the project root directory, I think that @roboLoki tried to build autotools. However, if do not set `ws2_32` as an additional link target, MinGW will not succeed.

MinGW users will need to build autotools according to the contents of `win32/INSTALL.txt`.

How about highlighting the introduction of `win32/INSTALL.txt` in the "Unix-like" section of `README.md`.

**Comment by jtv** [comment]
But the `configure` script adds `ws2_32` to the build when needed.  I wonder why that didn't work.

@tt4g I've just updated README.md to add a mention of win32/INSTALL.txt.  Thanks for that suggestion.

**Comment by tt4g** [comment]
Apparently the new MinGW does not inherit Windows environment variables by default.
Change the behavior by setting `MSYS2_PATH_TYPE`.

In the past, `MSYS2_PATH_TYPE=inherit` was specified, so may have found `ws2_32` that existed in the Windows `PATH` environment variable.

The recent MinGW seems to default to `MSYS2_PATH_TYPE=strict` and not inherit Windows environment variables. Because this can not find `ws2_32`, it may not be linked.

Sorry I can not investigate in detail because I do not have MinGW development environment.

**Comment by jtv** [comment]
Thanks for digging that up.  For now at least other people who hit the same problem have a chance to find that comment.

**Comment by gfgit** [comment]
Yes, I've commented out `operator==` and `operator<`.
Other operators like `operator!=` seem fine even if they call internally to the commented out operators.
Then I got some link errors in `utils.cxx` to `PQthreadsafe()` but after clean and rebuild they were gone.
Also as a note to documentation, when linking to `lipqxx`, I had to manually link also to `libpq` (of course) and on Windows to `ws2_32.dll`.
To ship the executable you also need some dlls found in PostgreSQL `bin` folder.
